### DISCUSSION TITLE
Option to use the app without secure-storage / biometrics

### DISCUSSION THREAD

**Discussion by vishnukvmd** [ente-io/ente#490] [Enhancements]
In offline mode, we generate a random `key` and encrypt the 2FA secrets with this key before persisting them to disk.

This `key` is then stored on the device's secure-storage, which [requires biometrics](https://github.com/ente-io/auth/blob/e1084c228fbb5b9f62f148b8747bf704a0f72a17/lib/onboarding/view/onboarding_page.dart#L185-L189) – that might be unavailable on some devices.

What we could do is provide an option to not use secure-storage, and instead encrypt the `key` with a passphrase, and prompt for this passphrase each time on app-start. Doing this will make the app accessible on devices that do not support biometric authentication.

**Comment by ghost** [comment]
Any updates?

**Comment by woowoo803** [comment]
Also interested in this. While I have biometrics on my phone it falls back to allowing my phone PIN so if someone observes me entering my PIN they can steal my phone and then can access all my TOTP codes. Having a strong password which needs to be entered each time would prevent this

**Reply by digitalfredy** [reply]
Will be great to have at least this option right now as workaround because in my case it just does not work.

Probably the cheap phone miss an android component related with biometrics (that enables the PIN option), if this is the case there is any app I cold install to get the biometrics dialog with falls back to PIN?

**Comment by vizzxc** [comment]
I am also interested in this feature. I opt not to use biometrics on my phone because I do not trust Samsung with that information. I would much rather prefer a pin or password feature to access my 2FA codes on the app.

**Comment by wiki-Bird** [comment]
> Also interested in this. While I have biometrics on my phone it falls back to allowing my phone PIN so if someone observes me entering my PIN they can steal my phone and then can access all my TOTP codes. Having a strong password which needs to be entered each time would prevent this


This feature is the only thing holding me back from switching to ente auth from authy. I will not use an authenticator that relies on the same passcode as my phone for the reason listed above.

**Comment by MagicFab** [comment]
This effectively prevents use on *some* Windows computers

**Comment by aooprojects** [comment]
I'm interested in this too and opt for the device PIN instead of biometrics with the following use case: I keep an old phone in a drawer as an backup to still have  OTP token(and online banking 2FA) when my regular phone breaks/is lost for some reason. I'm aware that I shouldn't use a phone with an really old Android version and ancient patch level, but this is just emergency backup and not for regular use!

**Comment by SofutoFukurou** [comment]
Is there any luck on this so far?

**Comment by brokeh** [comment]
I'm in exactly the same boat as @aooprojects - an old phone in the draw that is used as a backup.

I was able to resolve this by adding the following extra condition to https://github.com/ente-io/ente/blob/f0c935f835d514272a47d55528e2e7e2e87f00a7/mobile/apps/auth/lib/onboarding/view/onboarding_page.dart#L248-L251
```dart
|| await LocalAuthentication().isDeviceSupported()
```
As per the documentation in https://pub.dev/packages/local_auth#device-capabilities.

And then the only requirement is that the device has some sort of screen lock (e.g. pin), but doesn't actually need biometrics.

From what I can tell, that seems to work fine. I can add and import codes OK, they update correctly, etc. I assume they're still stored in the same way on the phone as if biometrics was used.

**Reply by AlcorSalvador** [reply]
All that would be missing is a warning. Otherwise, Nice!

**Reply by DefCon-CC** [reply]
> I'm in exactly the same boat as @aooprojects - an old phone in the draw that is used as a backup.
> 
> I was able to resolve this by adding the following extra condition to https://github.com/ente-io/ente/blob/f0c935f835d514272a47d55528e2e7e2e87f00a7/mobile/apps/auth/lib/onboarding/view/onboarding_page.dart#L248-L251
> 
> ```dart
> || await LocalAuthentication().isDeviceSupported()
> ```
> 
> As per the documentation in https://pub.dev/packages/local_auth#device-capabilities.
> 
> And then the only requirement is that the device has some sort of screen lock (e.g. pin), but doesn't actually need biometrics.
> 
> From what I can tell, that seems to work fine. I can add and import codes OK, they update correctly, etc. I assume they're still stored in the same way on the phone as if biometrics was used.

Can somebody create a PR for this?
